In [1]:
from google.colab import files

uploaded= files.upload()

Saving Online retail.csv to Online retail.csv


In [39]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [40]:
# Loading the dataset and checking size of the data

df = pd.read_csv("Online retail.csv", header=None)
df.shape

(7501, 1)

# **1. Data Preprocessing:**

In [41]:
df.head()

0
0  shrimp,almonds,avocado,vegetables mix,green gr...
1                             burgers,meatballs,eggs
2                                            chutney
3                                     turkey,avocado
4  mineral water,milk,energy bar,whole wheat rice...

In [42]:
# Converting data to appropriate format

df = df.join(df[0].str.split(',', expand=True).stack().unstack().reset_index(drop=True).rename(columns=lambda x: f'word_{x+1}'))

In [43]:
df.drop(0, axis=1, inplace=True)

In [44]:
df.head()

word_1     word_2      word_3            word_4        word_5  \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

             word_6 word_7          word_8        word_9       word_10  \
0  whole weat flour   yams  cottage cheese  energy drink  tomato juice   
1               NaN    NaN             NaN           NaN           NaN   
2               NaN    NaN             NaN           NaN           NaN   
3               NaN    NaN             NaN           NaN           NaN   
4               NaN    NaN             NaN           NaN           NaN   

          word_11    word_12 word_13 word_14        word_15 word_16  \
0  low fat yogurt  green tea   honey   salad  mineral water  salmon   
1             NaN        NaN     NaN     NaN            NaN     NaN   
2             NaN        NaN     NaN     NaN            NaN     NaN   
3             NaN        NaN     NaN     NaN            NaN     NaN   
4             NaN        NaN     NaN     NaN            NaN     NaN   

             word_17          word_18  word_19    word_20  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [45]:
df.shape

(7501, 20)

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 20 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   word_1   7501 non-null   object
 1   word_2   5747 non-null   object
 2   word_3   4389 non-null   object
 3   word_4   3345 non-null   object
 4   word_5   2529 non-null   object
 5   word_6   1864 non-null   object
 6   word_7   1369 non-null   object
 7   word_8   981 non-null    object
 8   word_9   654 non-null    object
 9   word_10  395 non-null    object
 10  word_11  256 non-null    object
 11  word_12  154 non-null    object
 12  word_13  87 non-null     object
 13  word_14  47 non-null     object
 14  word_15  25 non-null     object
 15  word_16  8 non-null      object
 16  word_17  4 non-null      object
 17  word_18  4 non-null      object
 18  word_19  3 non-null      object
 19  word_20  1 non-null      object
dtypes: object(20)
memory usage: 1.1+ MB


In [47]:
# Checking for missing values

df.isnull().sum()

word_1        0
word_2     1754
word_3     3112
word_4     4156
word_5     4972
word_6     5637
word_7     6132
word_8     6520
word_9     6847
word_10    7106
word_11    7245
word_12    7347
word_13    7414
word_14    7454
word_15    7476
word_16    7493
word_17    7497
word_18    7497
word_19    7498
word_20    7500
dtype: int64

# **2. Association Rule Mining:**

In [48]:
#pip install apyori

In [49]:
# Apriori package accepts only list so we will create an empty list and append all the transaction

from apyori import apriori


trans= []

for i in range(0, 7501):
    trans.append([str(df.values[i,j]) for j in range(0,20)])

In [50]:
trans

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers',
  'meatballs',
  'eggs',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['chutney',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['turkey',
  'avocado',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['mineral water',
  'milk',
  'energy bar',
  'whole wheat rice',
  'green tea',
  'nan',
  'nan',
  'nan',
 

In [51]:
# To get the rules from the below parameters

rules= apriori(transactions= trans,
              min_support=0.003,
              min_confidence= 0.2,
              min_lift= 3,
              min_length= 2, max_length= 2)
rules

<generator object apriori at 0x7e8fa862e3b0>

In [52]:
# Converting the rules into a list and saving it in results

results= list(rules)
results

[RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'pasta', 'escalope'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'fromage blanc', 'honey'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

In [53]:
# Checking for the length of results

len(results)

9

In [54]:
# 1st rule

results[0]

RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)])

In [55]:
# Creating empty list

base_item=[]
add_item=[]
support=[]
confidence=[]
lift=[]

In [56]:
# To access the required element from the results

results[0][2][0][0]   #Base item
results[0][2][0][1]   #Add item
results[0][1]         #Support
results[0][2][0][2]   #Confidence
results[0][2][0][3]   #Lift

4.84395061728395

In [57]:
# Using for loop to add all the required elements to the above empty list

for i in range(0,9):
    base_item.append(results[i][2][0][0])
    add_item.append(results[i][2][0][1])
    support.append(results[i][1])
    confidence.append(results[i][2][0][2])
    lift.append(results[i][2][0][3])

In [58]:
# Creating a dataframe of all the elements

pd.DataFrame(base_item)
pd.DataFrame(add_item)
pd.DataFrame(support)
pd.DataFrame(confidence)
pd.DataFrame(lift)

0
0  4.843951
1  3.790833
2  4.700812
3  5.164271
4  3.291994
5  3.840659
6  3.114710
7  4.122410
8  4.506672

In [59]:
# Concatenating all the above dataframe and saving it under new_data

new_data=pd.concat([pd.DataFrame(base_item),pd.DataFrame(add_item),pd.DataFrame(support),pd.DataFrame(confidence),pd.DataFrame(lift)], axis=1)

In [60]:
new_data.columns= ['Base item', 'Add item', 'Support', 'Confidence', 'Lift']
new_data

Base item     Add item   Support  Confidence      Lift
0           light cream      chicken  0.004533    0.290598  4.843951
1  mushroom cream sauce     escalope  0.005733    0.300699  3.790833
2                 pasta     escalope  0.005866    0.372881  4.700812
3         fromage blanc        honey  0.003333    0.245098  5.164271
4         herb & pepper  ground beef  0.015998    0.323450  3.291994
5          tomato sauce  ground beef  0.005333    0.377358  3.840659
6           light cream    olive oil  0.003200    0.205128  3.114710
7     whole wheat pasta    olive oil  0.007999    0.271493  4.122410
8                 pasta       shrimp  0.005066    0.322034  4.506672

# **3. Analysis and Interpretation:**

From the above data we can see that base item herb and pepper and add item ground beef has the highest support of 1.5 percent which means if someone purchases herb and pepper will also buy ground beef with a probability of 1.5%.

Moreover, base item tomato sauce and add item ground beef has a confidence of 37.7% which says that the person who buys tomato sauce will also buy ground beef with a probability of 37.7%

Furthermore, fromage blanc and honey has a lift of 5.16 which also means that fromage blanc is increasing the chance of buying honey 5.16 times.

1. What is lift and why is it important in Association rules?

> Lift measures how much more likely two items are to be bought together compared to how often they are bought individually.

It's important because it helps identify meaningful associations between items. A high lift value indicates a strong relationship between items, which can be useful for businesses to understand customer behavior and optimize their strategies, like product placement or marketing.

2. What is support and Confidence. How do you calculate them?

> Support: It refers to the frequency or occurence of a particular set of items in a dataset. It indicates how often a specific combination of items appears together in the data. Higher support values suggest stronger relationships or associations between the items, making them more significant for generating association rules.

Support is calculated by dividing the number of transactions containing the item set by the total number of transactions.

Confidence: It is the measure that shows how often two things (like items in a dataset) tend to occur together. It helps us understand the likelihood of one thing happening when the other occurs. Higher confidence means a stronger connection between two items.

Confidence is calculated by dividing the number of transactions containing both items in the rule by the number of transactions containing the first item.

confidence(A/B)= (A intersection B)/B (Conditional probability)

3.	What are some limitations or challenges of Association rules mining?

> One challenge is the presence of spurious or random associations, especially in large datasets. Not all associations identified are meaningful or actionable.

Association rules don't consider the order of item occurrence in transactions, which can limit their applicability in certain scenarios.

They may not capture complex relationships or dependencies between items, particularly when dealing with a large number of items.

Association rules mining requires a significant amount of data to generate reliable results, and the interpretation of rules often requires domain knowledge to be meaningful.